File: python/complex-numbers/complex_numbers.py
import cmath

ROUND_DIGIT_NUM = 15


class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    @classmethod
    def from_complex(cls, c):
        return ComplexNumber(round(c.real, ROUND_DIGIT_NUM), round(c.imag, ROUND_DIGIT_NUM))

    def __eq__(self, other):
        return self.real, self.imaginary == other.real, other.imaginary

    def __ne__(self, other):
        return self.real, self.imaginary != other.real, other.imaginary

    def __add__(self, other):
        return ComplexNumber.from_complex(complex(self.real, self.imaginary) + complex(other.real, other.imaginary))

    def __mul__(self, other):
        return ComplexNumber.from_complex(complex(self.real, self.imaginary) * complex(other.real, other.imaginary))

    def __sub__(self, other):
        return ComplexNumber.from_complex(complex(self.real, self.imaginary) - complex(other.real, other.imaginary))

    def __truediv__(self, other):
        return ComplexNumber.from_complex(complex(self.real, self.imaginary) / complex(other.real, other.imaginary))

    def __abs__(self):
        return abs(complex(self.real, self.imaginary))

    def conjugate(self):
        return ComplexNumber.from_complex(complex(self.real, self.imaginary).conjugate())

    def exp(self):
        return ComplexNumber.from_complex(cmath.exp(complex(self.real, self.imaginary)))

File: cpp/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"

namespace complex_numbers
{
Complex::Complex(const std::complex<double> &c) : c{c} {}

Complex::Complex(double real, double imag) : c{real, imag} {}

double Complex::real() const { return c.real(); }

double Complex::imag() const { return c.imag(); }

double Complex::abs() const { return std::abs(c); }

Complex Complex::conj() const { return std::conj(c); };

Complex Complex::exp() const { return std::exp(c); };

Complex operator+(const Complex &c1, const Complex &c2) { return c1.c + c2.c; }

Complex operator-(const Complex &c1, const Complex &c2) { return c1.c - c2.c; }

Complex operator*(const Complex &c1, const Complex &c2) { return c1.c * c2.c; }

Complex operator/(const Complex &c1, const Complex &c2) { return c1.c / c2.c; }
} // namespace complex_numbers
